src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+        {
+            changes.emplace_back(LEDGER_ENTRY_STATE);
+            changes.back().state() = *previous;
+            if (entry)
+            {
+                changes.emplace_back(LEDGER_ENTRY_UPDATED);
+                changes.back().updated() = *entry;
+            }
+            else
+            {
+                changes.emplace_back(LEDGER_ENTRY_REMOVED);
+                changes.back().removed() = state.first;
+            }
+        }
+        else
+        {
+            if (entry)
+            {
+                changes.emplace_back(LEDGER_ENTRY_CREATED);
+                changes.back().created() = *entry;
+            }
+        }
+    }
+    return changes;
+}
+
+std::vector<LedgerEntry>
+LedgerState::getLiveEntries()
+{
+    assert(mRoot);
+    assert(!mChild);
+
+    std::vector<LedgerEntry> entries;
+    for (auto const& state : mState)
+    {
+        auto le = state.second->ignoreInvalid().entry();
+        if (le)
+        {
+            entries.push_back(*le);
+        }
+    }
+    return entries;
+}
+
+std::vector<LedgerKey>
+LedgerState::getDeadEntries()
+{
+    assert(mRoot);
+    assert(!mChild);
+
+    std::vector<LedgerKey> keys;
+    for (auto const& state : mState)
+    {
+        if (state.second->ignoreInvalid().entry() &&
+            !state.second->ignoreInvalid().previousEntry())
+        {
+            keys.push_back(state.first);
+        }
+    }
+    return keys;
+}
+
+void
+LedgerState::commit(std::function<void()> f)
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    if (mParent)
+    {
+        assert(!f);
+        mergeStateIntoParent();
+        mergeHeaderIntoParent();
+        mParent->mChild = nullptr;
+        mParent = nullptr;
+    }
+    else
+    {
+        soci::transaction sqlTx(mRoot->getDatabase().getSession());
+        try
+        {
+            mergeStateIntoRoot();
+            mergeHeaderIntoRoot();
+            if (f)
+            {
+                f();
+            }
+            sqlTx.commit();
+        }
+        catch (...)
+        {
+            mRoot->getCache().clear();
+            throw;
+        }
+
+        mRoot->hasChild(false);
+        mRoot = nullptr;
+    }
+
+    mHeader.reset();
+    mState.clear();
+    mLoadBestOfferContext.clear();
+}
+
+void
+LedgerState::mergeStateIntoParent()
+{
+    for (auto const& state : mState)
+    {
+        state.second->invalidate();
+        auto const& entry = state.second->ignoreInvalid().entry();
+
+        auto iter = mParent->mState.find(state.first);
+        if (iter == mParent->mState.end())
+        {
+            StateEntry ler(new LedgerEntryReference(entry, nullptr));
+            ler->invalidate();
+            mParent->mState[state.first] = ler;
+        }
+        else
+        {
+            auto const& previous =
+                iter->second->ignoreInvalid().previousEntry();
+            StateEntry ler(new LedgerEntryReference(entry, previous));
+            ler->invalidate();
```

Comment:
This means that we don't want anyone to use that LedgerEntryReference anymore? Even if it was just created here and is 100% valid?

Do I understand this correctly? If so, please and comment and explain why is that needed. I think that this means we won't be able to use this entry in this LedgerState and one will have to create a new child LedgerState to use it?

Reply:
The `LedgerEntryReference` that is constructed here isn't exactly being created here. This is actually the process of copying the data from the child `LedgerState` to the parent `LedgerState`. So there was already a `LedgerEntryReference` that referenced this data, and it was already used as much as needed in the child `LedgerState` that is being committed.

I adopted the convention that any transaction-type operation (for example: commit, rollback, sub-transaction) would invalidate all existing `LedgerEntryReference`. Being invalid does not mean that you need to create a child `LedgerState` in order to use it, because it is valid to call `load` on the `LedgerKey` of an invalid `LedgerEntryReference` (see lines 372-397). Indeed, this is exactly what is done in the following example:
```
LedgerKey key = ...;
LedgerState lsParent(...);
LedgerState lsChild(lsParent);
someFunction(lsChild, key); // Some rarely accessed branches call lsChild.load(key)
lsChild.commit();
lsParent.load(key); 
```
Because the call to `lsChild.commit()` leaves every `LedgerEntryReference` in the invalid state, `lsParent.load(key)` will never fail with "Valid LedgerEntryReference already exists for this key" even if `someFunction` did not leave the `LedgerEntryReference` corresponding to `key` in the invalid state. 

It is worth noting that "re-loading" the invalidated `LedgerEntryReference` is a purely in-memory operation.

Let me know if this is still not clear, happy to explain more thoroughly or give more examples of different usage patterns (it might also be interesting to look at LedgerStateTests.cpp to see what kind of usage is possible).